cpp/Libusb.cpp

```
@@ -0,0 +1,138 @@
+}
+
+std::variant<std::monostate, libusb::Error> libusb::Endpoint::in_clear_halt() const {
+    int const r = libusb_clear_halt(inner->devh, ep_in);
+    if (r != 0) {
+        return libusb::Error(r);
+    }
+    return std::monostate();
+}
+
+std::variant<int, libusb::Error>  libusb::Endpoint::bulk_transfer(uint8_t ep, unsigned char *data, int length, unsigned int timeout) const {
+    int actual_length = 0;
+    int const r = libusb_bulk_transfer(inner->devh, ep, data, length, &actual_length, timeout);
+    if (r < 0) {
+        libusb::Error err(r);
+        switch (r) {
+        case LIBUSB_ERROR_TIMEOUT:
+            if (actual_length != 0 && actual_length != length && (ep_in & 0x80) != 0) {
+                //std::cerr << "Uh, oh. Transferred " << actual_length << " bytes on timeout" << std::endl;
+                return actual_length;
+            }
+            break;
+        default:
+            std::cerr << "Libusb: libusb_bulk_transfer failed: " << err.get_message() << std::endl;
+            break;
+        }
+        return err;
+    }
+    return actual_length;
+}
+
+std::variant<std::string, libusb::Error> libusb::Open_device::get_string_descriptor(uint8_t index) {
+    std::string ret(1024, '\0');
+    auto const r = libusb_get_string_descriptor_ascii(
+        inner->devh, index, reinterpret_cast<unsigned char *>(ret.data()), ret.size());
+    if (r < 0) {
+        libusb::Error err(r);
+        if (!inner->silent) {
+            std::cerr << "Libusb: Error getting string descriptor: " << err.get_message() << std::endl;
+        }
+        return err;
+    }
+    ret.resize(r);
+    return ret;
+}
+
+libusb::Claimed_interface::~Claimed_interface() {
+    if (inner == nullptr) {
+        return;
+    }
+    std::cout << "Libusb: libusb_release_interface" << std::endl;
+    int const r = libusb_release_interface(inner->devh, interface_number);
+    if (r != 0) {
+        std::cerr << "Libusb: libusb_release_interface failed: " << r << std::endl;
```

Don't print here. Ignore the error.